hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/router/async/AsyncForEachRun.java

```diff
@@ -0,0 +1,185 @@
+ * @param <R> the type of the final result after applying the thenApply function
+ * @see AsyncRun
+ * @see AsyncApplyFunction
+ * @see BiFunction
+ */
+public class AsyncForEachRun<I, T, R> implements AsyncRun<R> {
+
+  private boolean satisfy = false;
+  private Iterator<I> iterator;
+  private I now;
+  private final CompletableFuture<R> result = new CompletableFuture<>();
+  private AsyncApplyFunction<I, T> asyncFunction;
+  private BiFunction<AsyncForEachRun<I, T, R>, T, R> thenApply;
+
+  @Override
+  public void run() {
+    try {
+      doOnce(null);
+    } catch (IOException ioe) {
+      result.completeExceptionally(ioe);
+    }
+    setCurCompletableFuture(result);
+  }
+
+  /**
+   * Performs a single iteration of the asynchronous for-each operation.
+   *
+   * <p>This method is called to process each element of the iterator provided to
+   * the {@link AsyncForEachRun} constructor. It applies the asynchronous function to
+   * the current element, then applies the 'then' function to the result. If the
+   * 'satisfy' condition is met, the iteration is halted, and the current result is
+   * used to complete the future. This method is recursive, so it will continue to
+   * call itself for the next elements until the iterator is exhausted or the satisfy
+   * condition is true.</p>
+   *
+   * @param ret the initial or current result to be passed into the 'then' function
+   * @throws IOException if an I/O error occurs while applying the asynchronous function
+   * @see #forEach(Iterator)
+   * @see #asyncDo(AsyncApplyFunction)
+   * @see #then(BiFunction)
+   */
+  private void doOnce(R ret) throws IOException {
+    if (!iterator.hasNext()) {
+      result.complete(ret);
+      return;
+    }
+    now = iterator.next();
+    CompletableFuture<T> completableFuture = asyncFunction.async(now);
+    completableFuture.thenApply(t -> {
+      R r = null;
+      try {
+        r = thenApply.apply(AsyncForEachRun.this, t);
+      } catch (IOException e) {
+        result.completeExceptionally(new CompletionException(e));
+        return null;
+      }
+      if (satisfy) {
+        result.complete(r);
+        return null;
+      }
+      try {
+        doOnce(r);
+      } catch (IOException e) {
+        throw new CompletionException(e);
+      }
+      return null;
+    }).exceptionally(e ->
+        result.completeExceptionally(e));
+  }
+
+  /**
+   * Retrieves the current element being processed in the asynchronous for-each loop.
+   *
+   * <p>This method provides access to the element that is currently being
+   * operated on within the asynchronous iteration. It can be useful for
+   * inspection, logging, or other purposes that require knowledge of the
+   * current state of the iteration.</p>
+   *
+   * @return the current element of type {@code I} being processed in the iterator.
+   * @see #forEach(Iterator)
+   * @see #run()
+   */
+  public I getNow() {
```

Comment:
The name of this function is not matched with annotation OR some ambiguity here? my first feeling it will return timestamp, but implement one abstract type returned actually here.

Reply:
done